agent/resources/resources_linux.go

```diff
@@ -0,0 +1,120 @@
+
+package resources
+
+import (
+	"github.com/aws/amazon-ecs-agent/agent/api"
+	"github.com/aws/amazon-ecs-agent/agent/config"
+	"github.com/aws/amazon-ecs-agent/agent/resources/cgroup"
+	"github.com/cihub/seelog"
+	"github.com/pkg/errors"
+)
+
+// cgroupWrapper implements the Resource interface
+type cgroupWrapper struct {
+	control cgroup.Control
+}
+
+// New is used to return an object that implements the Resource interface
+func New() Resource {
+	return newResources(cgroup.New())
+}
+
+func newResources(control cgroup.Control) Resource {
+	return &cgroupWrapper{
+		control: control,
+	}
+}
+
+// Init is used to initialize the resource
+func (c *cgroupWrapper) Init() error {
+	return c.cgroupInit()
+}
+
+// Setup sets up the resource
+func (c *cgroupWrapper) Setup(task *api.Task) error {
+	return c.setupCgroup(task)
+}
+
+// Cleanup removes the resource
+func (c *cgroupWrapper) Cleanup(task *api.Task) error {
+	return c.cleanupCgroup(task)
+}
+
+// cgroupInit is used to create the root '/ecs/ cgroup
+func (c *cgroupWrapper) cgroupInit() error {
+	if c.control.Exists(config.DefaultTaskCgroupPrefix) {
+		seelog.Debugf("Cgroup at %s already exists, skipping creation", config.DefaultTaskCgroupPrefix)
+		return nil
+	}
+	return c.control.Init()
+}
+
+// setupCgroup is used to create the task cgroup
+func (c *cgroupWrapper) setupCgroup(task *api.Task) error {
+	cgroupRoot, err := task.BuildCgroupRoot()
+	if err != nil {
+		return errors.Wrapf(err, "resource: setup cgroup: unable to determine cgroup root for task: %s", task.Arn)
+	}
+
+	seelog.Debugf("Setting up cgroup at: %s for task: %s", cgroupRoot, task.Arn)
+
+	if c.control.Exists(cgroupRoot) {
+		seelog.Debugf("Cgroup at %s already exists, skipping creation", cgroupRoot)
+		return nil
+	}
+
+	linuxResourceSpec, err := task.BuildLinuxResourceSpec()
+	if err != nil {
+		return errors.Wrapf(err, "resource: setup cgroup: unable to build resource spec for task: %s", task.Arn)
+	}
+
+	cgroupSpec := cgroup.Spec{
+		Root:  cgroupRoot,
+		Specs: &linuxResourceSpec,
+	}
+
+	cgrp, err := c.control.Create(&cgroupSpec)
+	if err != nil {
+		return errors.Wrapf(err, "resource: setup cgroup: unable to create cgroup at %s for task: %s", cgroupRoot, task.Arn)
+	}
+
+	// NOTE: This should be impossible
+	// cgrp is an interface and it should never be nil when err == nil
+	if cgrp == nil {
+		seelog.Criticalf("Invalid cgroup creation at %s", cgroupRoot)
+		return errors.Errorf("resource: setup cgroup: invalid cgroup object created for task: %s", task.Arn)
+	}
+
+	return nil
+}
+
+// cleanupCgroup is used to remove the task cgroup
+func (c *cgroupWrapper) cleanupCgroup(task *api.Task) error {
+	cgroupRoot, err := task.BuildCgroupRoot()
+	if err != nil {
+		return errors.Wrapf(err, "resource: cleanup cgroup: unable to determine cgroup root for task: %s", task.Arn)
+	}
+
+	seelog.Debugf("Cleaning up cgroup at: %s for task: %s", cgroupRoot, task.Arn)
+
+	err = c.control.Remove(cgroupRoot)
+	if err != nil {
```

Comment:
Please modify this to handle `ErrCgroupDeleted` (Ignore and log the error in such a case)

Reply:
Sure, I can handle this explicitly here but its hard to think of how we might end up with such a situation.

Reply:
We have seen the Agent trying to clean up the same task twice before because of bugs. No reason why we won't run into it again.